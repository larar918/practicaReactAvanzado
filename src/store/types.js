export const AUTH_LOGIN = 'auth/login';
export const AUTH_LOGOUT = 'auth/logout';

export const ADS_LOADED = 'ads/loaded'
export const ADS_CREATED = 'ads/created'
export const ADS_DELETED = 'ads/deleted'
