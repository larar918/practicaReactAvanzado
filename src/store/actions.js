// Funciones que generan acciones. Actions creators.

import { AUTH_LOGIN, AUTH_LOGOUT, ADS_LOADED,  ADS_DELETED, ADS_CREATED } from "./types"


export const authLogin = () => ({
    type: AUTH_LOGIN
});

export const authLogout = () => ({
    type: AUTH_LOGOUT
});


export const adsLoaded = (ads) => ({
    type: ADS_LOADED,
    payload: ads
});

export const adDeleted = (ad) => ({
    type: ADS_DELETED,
    payload: ad
});

export const adCreated = (ad) => ({
    type: ADS_CREATED,
    payload: ad,
});




