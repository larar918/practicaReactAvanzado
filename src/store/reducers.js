import {AUTH_LOGIN, AUTH_LOGOUT, ADS_CREATED, ADS_LOADED, ADS_DELETED} from './types.js'

const defaultState = {
    auth: false,
    ads : []
}

export function auth(state=defaultState.auth, action){  // Recibe como parámetro del estado solo state.auth y es con lo que trabaja
    switch(action.type){
        case AUTH_LOGIN:
            return true;
        case AUTH_LOGOUT:
            return false;
        default:
            return state;
    }
}

export function ads(state=defaultState.ads, action){ // Recibe como parámetro del estado solo state.ads y es con lo que trabaja
    switch(action.type){
        case ADS_LOADED:
            return action.payload;
        case ADS_DELETED:
            return state.filter(ad => ad.id !== action.payload); // Creamos un array nuevo sin el anuncio eliminado
        case ADS_CREATED:
            return [...state, action.payload]; // Añadimos el ad creado al listado
        default:
            return state;
    }
}

