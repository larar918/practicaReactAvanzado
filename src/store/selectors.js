export const getAds = state => state.ads

export const getAd = adId => state => getAds(state).find(ad => ad.id === adId)