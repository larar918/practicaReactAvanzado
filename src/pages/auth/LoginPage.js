import { login } from "./service";
import { useState } from "react";
import { useNavigate } from "react-router-dom";
import { useDispatch } from "react-redux";
import { authLogin } from "../../store/actions";

import Layout from "../../components/Layout";
import '../../styles/components.css';
import '../../styles/forms.css';


function LoginPage() {
  
  const dispatch = useDispatch();
  const navigate= useNavigate();
  
  const onLogin = () => {
    dispatch(authLogin(credentials));
  }

  // Estados para controlar si el botón de submit está o no  
  const [credentials, setCredentials] = useState({
    email: '',
    password:'',
    rememberMe: ''
  });
  
  const handleSubmit = async event => {
    event.preventDefault();
    await login(credentials);
    onLogin();
    navigate('/adverts'); 
  }

  // Estados para controlar los valores de email, user y checked. Creamos una manejador general el cual va a modificar solo la propiedad de
  // credentials que genera el cambio de estado (email/password/checked). Le pisamos la propiedad.
  const handleChange = event => {
    setCredentials(currentCredentials => ({
      ...currentCredentials,
      [event.target.name]: event.target.value,
    }));
  }

  // Constante para calcular en cada render (cada vez que se escribe un caracter en alguno de los inputs) si el botón tiene que seguir disabled
  const {email, password, checked} = credentials;
  const disabled = !(email && password); // si no hay email o contraseña su valor es true

  return (
    <Layout>
      <div className="page">
        <form onSubmit={ handleSubmit }>
          <fieldset>
              <h2 className="">Email: </h2><input data-testid="email" type="text" label="email" name="email" value = {email} onChange = {handleChange} className="formElement" /><br></br>
              <h2>Contraseña: </h2><input data-testid="password" type="password" label="password" name="password" value = {password} onChange = {handleChange}  className="formElement"/><br></br>
              <div className="checkElement">
                <input type="checkbox" name ="rememberMe" checked = {checked} onChange = {handleChange} className="formCheck" /> <p className="formatCheckTarget">Remember me!</p>
              </div> 
          </fieldset>
          <button type="submit" disabled = {disabled} className="submit disabled" >Log In</button>
        </form>
      </div>
    </Layout>
  );
}

export default LoginPage;




