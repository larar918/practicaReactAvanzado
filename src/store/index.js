import  { combineReducers, createStore } from 'redux';
import { devToolsEnhancer } from '@redux-devtools/extension';
import * as actionCreators from './actions'
import * as reducers from './reducers'

export default function configureStore(preloadedState){
    const store = createStore(
        combineReducers(reducers),
        preloadedState,
        devToolsEnhancer({ actionCreators }), // Uso de las devtools
    )
    return store;
}