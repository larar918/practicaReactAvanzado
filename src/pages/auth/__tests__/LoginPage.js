import { act, render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { authLogin } from '../../../store/actions';
import LoginPage from '../LoginPage';
import userEvent from '@testing-library/user-event';

const userType = (input, text) => userEvent.type(input, text);

jest.mock('../../../store/actions')

describe('LoginPage', () => {
  const state = {};
  const store = {
    getState: () => state,
    subscribe: () => {},
    dispatch: () => {},
  };

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <Router>
          <LoginPage />
        </Router>
      </Provider>
    );

  test('snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('should dispatch authLogin action', async () => {
    const email = 'keepcoder';
    const password = 'password';
    renderComponent();
  
    const emailInput = screen.getByTestId('email');
    const passwordInput = screen.getByTestId('password');
    const submitButton = screen.getByRole('button');
  
    console.log(document.body.innerHTML);  // Imprime el HTML del contenedor de prueba
  
    expect(submitButton).toBeDisabled();
    await act(() => userType(emailInput, email));
    await act(() => userType(passwordInput, password));
    expect(submitButton).toBeEnabled();
    await userEvent.click(submitButton);
   
    expect(authLogin).toHaveBeenCalledWith({ email, password });

  });
  
});
