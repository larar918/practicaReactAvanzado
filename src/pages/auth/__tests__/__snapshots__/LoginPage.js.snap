// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="layout"
  >
    <header>
      <div>
        <nav
          class="navbar"
        >
          <a
            class="navbarContent"
            href="/adverts"
          >
            NodePop
          </a>
          <div
            class="options"
          >
            <div
              class="left-options"
            />
          </div>
        </nav>
      </div>
    </header>
    <main>
      <div
        class="page"
      >
        <form>
          <fieldset>
            <h2
              class=""
            >
              Email: 
            </h2>
            <input
              class="formElement"
              data-testid="email"
              label="email"
              name="email"
              type="text"
              value=""
            />
            <br />
            <h2>
              Contrase√±a: 
            </h2>
            <input
              class="formElement"
              data-testid="password"
              label="password"
              name="password"
              type="password"
              value=""
            />
            <br />
            <div
              class="checkElement"
            >
              <input
                class="formCheck"
                name="rememberMe"
                type="checkbox"
              />
               
              <p
                class="formatCheckTarget"
              >
                Remember me!
              </p>
            </div>
          </fieldset>
          <button
            class="submit disabled"
            disabled=""
            type="submit"
          >
            Log In
          </button>
        </form>
      </div>
    </main>
  </div>
</div>
`;
